---
title: "Documentation of data analysis"
output:
  github_document:
    toc: true
    includes:
      before_body: header.md
    toc_depth: 4
  highlight: github
link-citations: true
always_allow_html: true
---

*** 
### Differential gene expression analysis using `DESeq2`   

The following pipeline was used to analyze RNA-seq data:

1. Raw sequencing reads in fastq format were filtered and trimmed for quality using fastp v. 0.23.2 to remove low-quality bases and adapter sequences (--cut_front --cut_tail -q 30).  

2. Ribosomal RNA reads were removed using sortmeRNA v. 4.3.6 based on sequences in the SILVA database.  
3. The rRNA-depleted reads were aligned to the Pyrococcus furiosus DSM 3638 reference genome (NCBI: CP023154.1) using Bowtie2 v. 2.5.0 with default parameters.  

4. The resulting sequence alignment files (SAM) were converted to binary mapping format (BAM) using samtools.  

5. Differentially expressed genes were identified using DESeq2 package following the recommendations in the Bioconductor vignette.  

6. FeatureCounts from RSubread package v. 2.10.5 was used to calculate the count matrix based on a custom GTF file generated by filtering the P. furiosus DSM 3638 GFF annotation file downloaded from NCBI for protein-coding genes (column biotype).  

7. Principal component analysis (PCA) was performed on variance stabilizing transformed (VST) data, and outlier replicates were removed from the dataset after visual inspection.  

8. Differential expression analysis was conducted by comparing each of the cold or heat shock conditions with the control condition in a pairwise manner.  


### Identification of enriched 3' ends from Term-seq data  

### Coverage analysis of long Nanopore reads 

### Differential protein expression analysis using `limma` 

### Downstream data analysis 
#### arCOG enrichment analyis  
#### Promoter and terminator analysis  
#### Operon coverage analysis 
#### Correlation analysis  
#### Detection and analysis of signature clusters  


### Tools used 

- fastp v. 0.23.2: https://github.com/OpenGene/fastp 
- sortmeRNA v. 4.3.6: https://github.com/biocore/sortmerna 
- Bowtie2 v. 2.5.0: https://github.com/BenLangmead/bowtie2 
- samtools: https://github.com/samtools/samtools 
- DESeq2 package: https://bioconductor.org/packages/release/bioc/html/DESeq2.html 
- RSubread package v. 2.10.5: https://bioconductor.org/packages/release/bioc/html/Rsubread.html 



