---
title: "Documentation of data analysis"
output:
  github_document:
    toc: true
    includes:
      before_body: header.md
    toc_depth: 4
  highlight: github
link-citations: true
always_allow_html: true
---

*** 
### Differential gene expression analysis using `DESeq2`   

The following pipeline was used to analyze RNA-seq data:

1. Raw sequencing reads in fastq format were filtered and trimmed for quality using fastp v. 0.23.2 to remove low-quality bases and adapter sequences (--cut_front --cut_tail -q 30).  

2. Ribosomal RNA reads were removed using sortmeRNA v. 4.3.6 based on sequences in the SILVA database.  
3. The rRNA-depleted reads were aligned to the *Pyrococcus furiosus* DSM 3638 reference genome (NCBI: CP023154.1) using Bowtie2 v. 2.5.0 with default parameters.  

4. The resulting sequence alignment files (SAM) were converted to binary mapping format (BAM) using samtools.  

5. Differentially expressed genes were identified using DESeq2 package following the recommendations in the Bioconductor vignette.  

6. FeatureCounts from RSubread package v. 2.10.5 was used to calculate the count matrix based on a custom GTF file generated by filtering the *P. furiosus* DSM 3638 GFF annotation file downloaded from NCBI for protein-coding genes (column biotype).  

7. Principal component analysis (PCA) was performed on variance stabilizing transformed (VST) data, and outlier replicates were removed from the dataset after visual inspection.  

8. Differential expression analysis was conducted by comparing each of the cold or heat shock conditions with the control condition in a pairwise manner.  


### Identification of enriched 3' ends from Term-seq data  

1. Quality and adapter trimming of the raw fastq files was performed using trimmomatic in paired-end mode, with the following settings: ILLUMINACLIP:TruSeq3-PE.fa:2:30:10:8:true, AVGQUAL:25, MINLEN:20. 

2. Unique molecular identifiers (UMIs) were extracted from the paired-end reads using the umi_tools command (--bcpattern=NNNN, --bc-pattern2=NNNN).  

3. Mapping of the reads to the reference genome was performed using bowtie2 in --sensitive-local mode with the last four nucleotides of each read being trimmed using --trim3 4. SAM files were then converted to BAMs, sorted and indexed using samtools.  

4. Deduplication of mapped reads was performed using the extracted UMIs by umi_tools dedup with default settings for paired-end data. 

5. Detection of enriched 3' ends by peak calling and downstream analysis was performed as follows: 
    a. Strand-specific bedgraph files were generated and CPM normalized using deepTools bamCoverage, with the SAM flags 83 and 99 in combination with --Offset 1 and --binSize 1 --minMappingQuality 20. 
    b. The termseq_peaks script was used with default parameters to call peaks using four replicates as input. This script can be found on GitHub at https://github.com/nichd-bspc/termseq-peaks. 
    c. For end detection, peak files were merged with the position-specific count files using bedtools intersect (option -wao). 
    d. Enriched positions were filtered and annotated based on the following criteria: For each peak, the position with the highest number of reads was selected per replicate and only maximum peak positions selected that were present in at least three of the four replicates. Positions with less than five CPM counts were excluded from the analysis. Positions were assigned based on their relative orientation to a gene and their respective peak height as primary (highest peak within 300 bases downstream from a gene), secondary (each additional peak 300 bases downstream from a gene), and internal (each peak in the coding range). 
  

### Coverage analysis of long Nanopore reads  
1. Basecalling of fast5 files was performed using guppy (v. 6.4.2+97a7f06) in high-accuracy mode (dna_r9.4.1_450bpd_hac.cfg) with standard parameters and a quality score threshold of 9.  

2. The PCR-cDNA library was demultiplexed by guppy_barcoder using default settings (--barcode_kits SQK-PCB109), except for disabling barcode trimming.  

3. Full-length sequenced reads were identified, strand-oriented, and trimmed using pychopper (v. 2.7.2, https://github.com/epi2me-labs/pychopper) with autotuned cutoffs and the recommended edlib backend for identifying custom primers.  

4. Cutadapt (v. 4.2) was used to remove remaining 3'-adapter sequences with the parameters -a "CTGTAGGCACCATCAAT" -j 0.  

5. Next, trimmed reads were mapped using minimap2 (v. 2.24-r1122) with standard parameters suggested for aligning Nanopore genomic reads (-ax map-ont) to the *P. furiosus* DSM 3638 genome (NCBI: CP023154.1). Alignments with more than 5 clipped bases (soft or hard clips) were removed using samclip (v. 0.4.0), and SAM files were converted to sorted BAM files using samtools (v. 1.16.1).  

6. Lastly, strand-specific coverage files were created using samtools depth (-a, -J options enabled) with a binsize of 1 and including reads with deletions in the coverage computation. Downstream analysis, including CPM normalization, calculating the mean coverage for each position of the two replicates, and plotting, was performed using the Tidyverse in R.



### Differential protein expression analysis using `limma`  
1. Quantitative values were log2-transformed and subjected to PCA for quality control and detection of outlier replicates (CS 2 replicate 1, CS 3 replicate 4, CS R replicate 2, HS 2 replicate 3, HS R replicate 1) after visual inspection.  

2. Normalized protein expression data was analyzed using the R limma (v. 3.52.4) package to identify differentially expressed proteins between control and cold or heat-stressed samples. 

### Downstream data analysis 
#### Normalisation  
For RNA-seq data analysis, total transcripts per million (TPM) was used as a normalization method to adjust for differences in sequencing depth and transcript length. TPM values were calculated using the following formula: 

    (number of reads mapping to each gene / gene length in kilobases) * (1,000,000 / sum of all RPK values in the sample)

This method ensures that the sum of all TPM values in each sample is the same, facilitating comparison of gene expression levels between samples. 


#### arCOG enrichment analyis  
1. For functional enrichment analysis based on the Archaeal Clusters of Orthologous Genes (arCOG), arCOGs for *P. furiosus* were retrieved from  ftp://ftp.ncbi.nih.gov/pub/wolf/COGs/arCOG/ 

2. Gene set enrichment analysis performed with the [goseq](https://bioconductor.org/packages/release/bioc/html/goseq.html) (v. 1.48.0) package in R, which accounts for gene length bias. For each comparison, a condition- and method-specific background file was generated from all detectable genes. Next, p-values for overrepresentation of arCOG terms in the differentially expressed genes were separately calculated for up- and downregulated genes based on RNA-seq and MS data, respectively. Significantly enriched terms were identified with a cutoff of 0.05.  

#### Promoter and terminator analysis  
##### 5' ends  
1. Primary transcription start sites (TSS) and corresponding 5' UTR lengths were extracted from the supplementary table in https://www.frontiersin.org/articles/10.3389/fmicb.2019.01603/full

2. Genes were categorized based on the presence of an archaeal-typical promoter motif containing a TFB-recognition element (BRE) and a TATA element after MEME motif search (sequences from -50 to +10 from the TSS analyzed via MEME (v. 5.4.1., -mod zoops -minw 8 -maxw 20)  

3. Position-specific motifs plotted in R using the ggseqlogo (v. 0.1) package  

4. Promoter strength was estimated according to the method used in https://pubmed.ncbi.nlm.nih.gov/34535658/ by analyzing sequences from -42 to -19 using MEME in oops mode. The p-value from the motif search used as an estimator for the promoter strength 

##### 3' ends 
1. 3' ends classified similarly based on the presence of a poly(U)-terminator motif. Sequences from -35 to +2 from the primary 3' ends extracted and analyzed via MEME (-mod zoops -minw 4 -maxw 20) and genes contributing to the motif categorized as +poly(U) motif 

2. Nucleotide enrichment analysis performed as described in https://pubmed.ncbi.nlm.nih.gov/27670118/, comparing the frequency of each base calculated using the extracted sequences to the same calculation based on randomly sampling 100,000 positions from intergenic regions, and only the enrichment position of nucleotide U plotted

3. Structural stability of the RNA predicted by folding the 45 nt long RNA upstream of the terminators using the LncFinder (v. 1.1.5) package in R.  

#### Correlation analysis  
The Pearson correlation coefficient was used to calculate the correlation between log2-fold changes and count values found in both samples (pairwise complete). 

#### Detection and analysis of signature clusters  
Principal component analysis (PCA) was performed on the z-score normalized log2 fold changes of RNA and protein values using the prcomp function in R.  
1. Pairwise Euclidean distances between samples were calculated from the PCA results and used to perform hierarchical clustering with the ward.D2 linkage method. 

2. The number of clusters was determined using the elbow method and enrichment analysis of arCOG categories. This allowed identification of clusters of genes with similar expression patterns across different conditions and datasets, providing insights into the regulation of cellular processes in Pyrococcus furiosus.

For downstream analysis: 
- The codon adaptation index (CAI) was computed using the CAI function from the coRdon package. The 5% most abundant proteins according to our control sample were used for the computation.
The codon adaptation index (CAI) is a measure of the similarity between the codon usage of a gene and the codon usage of highly expressed genes in the organism.
- Known and predicted protein-protein interactions (PPIs) among the differentially expressed genes were identified using the STRING database.  

### Tools used 

- fastp v. 0.23.2: https://github.com/OpenGene/fastp 
- sortmeRNA v. 4.3.6: https://github.com/biocore/sortmerna 
- Bowtie2 v. 2.5.0: https://github.com/BenLangmead/bowtie2 
- samtools: https://github.com/samtools/samtools 
- DESeq2 v. 1.36.0: https://bioconductor.org/packages/release/bioc/html/DESeq2.html 
- RSubread package v. 2.10.5: https://bioconductor.org/packages/release/bioc/html/Rsubread.html 
- Trimmomatic v. 0.39: https://github.com/usadellab/Trimmomatic 
- UMI-tools v.1.0.1: https://github.com/CGATOxford/UMI-tools 
- deepTools v. 3.5.0: https://deeptools.readthedocs.io 
- Termseq-peaks: https://github.com/nichd-bspc/termseq-peaks 
- guppy v. 6.4.2+97a7f06
- pychopper v. 2.7.2: https://github.com/epi2me-labs/pychopper 
- cutadapt v. 4.2: https://github.com/marcelm/cutadapt
- minimap2 v. 2.24-r1122: https://github.com/lh3/minimap2
